File: web/src/core/curve.ts
/**
 * Monotone cubic (Fritsch-Carlson) curve — mirror of `pw_color/curve.py`.
 *
 * See that file for why this is not Catmull-Rom. Short version: Catmull-Rom
 * overshoots and can reverse, which on a tone curve means a brighter input
 * mapping to a darker output. Fritsch-Carlson cannot do that.
 *
 * `CurvePoint` is `[x, y]` rather than `{x, y}` so that the control points
 * serialise into the workflow JSON as compact arrays.
 */

export type CurvePoint = [number, number];

export const IDENTITY_POINTS: CurvePoint[] = [
  [0, 0],
  [1, 1],
];

export function isIdentity(points: CurvePoint[]): boolean {
  if (points.length !== 2) return false;
  const [[x0, y0], [x1, y1]] = points;
  return Math.abs(x0) < 1e-9 && Math.abs(y0) < 1e-9 && Math.abs(x1 - 1) < 1e-9 && Math.abs(y1 - 1) < 1e-9;
}

/**
 * Sort by x and collapse duplicate x, keeping the last. Dragging a point past
 * its neighbour is a normal mid-gesture state, so this is handled here rather
 * than forbidden in the editor.
 */
function sortedUnique(points: CurvePoint[]): { xs: number[]; ys: number[] } {
  const pts = points.map((p) => [p[0], p[1]] as CurvePoint).sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  const xs: number[] = [];
  const ys: number[] = [];
  for (const [x, y] of pts) {
    if (xs.length && Math.abs(x - xs[xs.length - 1]) < 1e-7) ys[ys.length - 1] = y;
    else {
      xs.push(x);
      ys.push(y);
    }
  }
  if (xs.length < 2) {
    const y = ys.length ? ys[0] : 0;
    return { xs: [0, 1], ys: [y, ys.length ? y : 1] };
  }
  return { xs, ys };
}

export function monotoneTangents(xs: number[], ys: number[]): number[] {
  const n = xs.length;
  const h: number[] = [];
  const delta: number[] = [];
  for (let i = 0; i < n - 1; i++) {
    h.push(xs[i + 1] - xs[i]);
    delta.push((ys[i + 1] - ys[i]) / h[i]);
  }

  const m = new Array<number>(n).fill(0);
  m[0] = delta[0];
  m[n - 1] = delta[n - 2];
  for (let i = 1; i < n - 1; i++) {
    if (delta[i - 1] * delta[i] <= 0) {
      m[i] = 0; // local extremum — the flat tangent is what kills the overshoot
    } else {
      const w1 = 2 * h[i] + h[i - 1];
      const w2 = h[i] + 2 * h[i - 1];
      m[i] = (w1 + w2) / (w1 / delta[i - 1] + w2 / delta[i]);
    }
  }

  // Fritsch-Carlson circle condition — also constrains the endpoint tangents,
  // which the weighted mean above never touches.
  for (let i = 0; i < n - 1; i++) {
    if (delta[i] === 0) {
      m[i] = 0;
      m[i + 1] = 0;
      continue;
    }
    const a = m[i] / delta[i];
    const b = m[i + 1] / delta[i];
    const s = a * a + b * b;
    if (s > 9) {
      const t = 3 / Math.sqrt(s);
      m[i] = t * a * delta[i];
      m[i + 1] = t * b * delta[i];
    }
  }
  return m;
}

/**
 * A prepared curve. Building the tangents once and evaluating many times is
 * what keeps a 35 937-point lattice build off the frame budget.
 */
export class Curve {
  readonly xs: number[];
  readonly ys: number[];
  readonly m: number[];

  constructor(points: CurvePoint[]) {
    const { xs, ys } = sortedUnique(points);
    this.xs = xs;
    this.ys = ys;
    this.m = monotoneTangents(xs, ys);
  }

  /** Evaluate at x, linearly extended past the ends, clamped to [0,1]. */
  at(x: number): number {
    const { xs, ys, m } = this;
    const n = xs.length;
    if (x < xs[0]) return clamp01(ys[0] + (x - xs[0]) * m[0]);
    if (x > xs[n - 1]) return clamp01(ys[n - 1] + (x - xs[n - 1]) * m[n - 1]);

    let lo = 0;
    let hi = n - 1;
    while (hi - lo > 1) {
      const mid = (lo + hi) >> 1;
      if (xs[mid] <= x) lo = mid;
      else hi = mid;
    }
    const h = xs[lo + 1] - xs[lo];
    const t = (x - xs[lo]) / h;
    const t2 = t * t;
    const t3 = t2 * t;
    const h00 = 2 * t3 - 3 * t2 + 1;
    const h10 = t3 - 2 * t2 + t;
    const h01 = -2 * t3 + 3 * t2;
    const h11 = t3 - t2;
    return clamp01(h00 * ys[lo] + h10 * h * m[lo] + h01 * ys[lo + 1] + h11 * h * m[lo + 1]);
  }
}

function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}

File: pw_color/curve.py
"""Monotonic cubic interpolation for the curve editor.

Fritsch-Carlson (1980), not Catmull-Rom. Catmull-Rom is the usual choice and it
is wrong for tone curves: with unevenly spaced control points it overshoots,
which shows up as a dark halo just above a lifted shadow point and, worse, as a
*reversal* — a range of inputs where a brighter input maps to a darker output.
Every existing ComfyUI curve node has this. Fritsch-Carlson limits the tangents
so the interpolant is monotone wherever the control points are, which makes
overshoot and reversal impossible by construction rather than by clamping after
the fact.

Reference: Fritsch & Carlson, "Monotone Piecewise Cubic Interpolation",
SIAM J. Numer. Anal. 17(2), 1980.

``web/src/core/curve.ts`` mirrors this file exactly. If you change the maths
here, change it there, and the parity test will tell you if you didn't.
"""

from __future__ import annotations

import torch

__all__ = ["monotone_tangents", "eval_curve", "IDENTITY_POINTS"]

#: A two-point identity curve. The minimum a curve is ever allowed to be.
IDENTITY_POINTS: tuple[tuple[float, float], ...] = ((0.0, 0.0), (1.0, 1.0))


def _sorted_unique(points: list[tuple[float, float]]) -> tuple[list[float], list[float]]:
    """Sort by x and drop duplicate x, keeping the last.

    Duplicate x values would divide by zero in the slope. Dragging a point past
    its neighbour is a normal thing for a user to do mid-gesture, so this has to
    be handled here rather than prevented in the UI.
    """
    pts = sorted((float(x), float(y)) for x, y in points)
    xs: list[float] = []
    ys: list[float] = []
    for x, y in pts:
        if xs and abs(x - xs[-1]) < 1e-7:
            ys[-1] = y
        else:
            xs.append(x)
            ys.append(y)
    if len(xs) < 2:
        return [0.0, 1.0], [ys[0] if ys else 0.0, ys[0] if ys else 1.0]
    return xs, ys


def monotone_tangents(xs: list[float], ys: list[float]) -> list[float]:
    """Fritsch-Carlson tangents ``m[i]`` at each knot."""
    n = len(xs)
    h = [xs[i + 1] - xs[i] for i in range(n - 1)]
    delta = [(ys[i + 1] - ys[i]) / h[i] for i in range(n - 1)]

    m = [0.0] * n
    m[0] = delta[0]
    m[n - 1] = delta[n - 2]
    for i in range(1, n - 1):
        if delta[i - 1] * delta[i] <= 0.0:
            # Local extremum: flat tangent, which is what stops the overshoot.
            m[i] = 0.0
        else:
            # Weighted harmonic mean (Fritsch-Butland form): favours the
            # shorter interval, so a tight pair of points doesn't get whipped
            # around by a distant one.
            w1 = 2.0 * h[i] + h[i - 1]
            w2 = h[i] + 2.0 * h[i - 1]
            m[i] = (w1 + w2) / (w1 / delta[i - 1] + w2 / delta[i])

    # Fritsch-Carlson circle condition. Belt and braces over the harmonic mean:
    # it also handles the endpoint tangents, which the mean never touches.
    for i in range(n - 1):
        if delta[i] == 0.0:
            m[i] = 0.0
            m[i + 1] = 0.0
            continue
        a = m[i] / delta[i]
        b = m[i + 1] / delta[i]
        s = a * a + b * b
        if s > 9.0:
            t = 3.0 / (s**0.5)
            m[i] = t * a * delta[i]
            m[i + 1] = t * b * delta[i]
    return m


def eval_curve(points: list[tuple[float, float]], x: torch.Tensor) -> torch.Tensor:
    """Evaluate the monotone cubic through ``points`` at ``x``.

    Outside the control point range the curve is extended linearly along the
    end tangent, then the whole thing is clamped to ``[0,1]`` — a tone curve
    that leaves the display range is not doing anything a user asked for.
    """
    xs, ys = _sorted_unique(list(points))
    m = monotone_tangents(xs, ys)
    n = len(xs)

    # Preserve the caller's dtype. Lattice bakes run in float64 so that the
    # result matches the browser's float64 arithmetic bit for bit after
    # quantisation; forcing float32 here would reintroduce the drift.
    dev = x.device
    dt = x.dtype if x.dtype.is_floating_point else torch.float32
    xt = torch.tensor(xs, dtype=dt, device=dev)
    yt = torch.tensor(ys, dtype=dt, device=dev)
    mt = torch.tensor(m, dtype=dt, device=dev)

    xf = x.to(dt)
    # Segment index: the last knot whose x is <= our x.
    idx = (torch.bucketize(xf.contiguous(), xt, right=True) - 1).clamp(0, n - 2)

    x0, x1 = xt[idx], xt[idx + 1]
    y0, y1 = yt[idx], yt[idx + 1]
    m0, m1 = mt[idx], mt[idx + 1]
    h = x1 - x0
    t = (xf - x0) / h

    t2 = t * t
    t3 = t2 * t
    h00 = 2.0 * t3 - 3.0 * t2 + 1.0
    h10 = t3 - 2.0 * t2 + t
    h01 = -2.0 * t3 + 3.0 * t2
    h11 = t3 - t2
    y = h00 * y0 + h10 * h * m0 + h01 * y1 + h11 * h * m1

    # Linear extension beyond the endpoints, using the end tangents.
    below = xf < xt[0]
    above = xf > xt[n - 1]
    y = torch.where(below, yt[0] + (xf - xt[0]) * mt[0], y)
    y = torch.where(above, yt[n - 1] + (xf - xt[n - 1]) * mt[n - 1], y)
    return y.clamp(0.0, 1.0)
